### PR TITLE
Update to secp256k1 0.21.2

### PR DESCRIPTION
**Pull request by sanket1729**

### CODE REVIEW

**Comment by sanket1729** [comment]
This is ready for review now and kind of a blocker for the rest of the PRs for the next release. @apoelstra, @dr-orlovsky , @Kixunil

File: src/util/taproot.rs [hunk 1]
```diff
@@ -700,8 +700,7 @@ impl ControlBlock {
 
     /// Serialize to a writer. Returns the number of bytes written
     pub fn encode<Write: io::Write>(&self, mut writer: Write) -> io::Result<usize> {
-        let first_byte: u8 =
-            (if self.output_key_parity { 1 } else { 0 }) | self.leaf_version.as_u8();
+        let first_byte: u8 = i32::from(self.output_key_parity) as u8 | self.leaf_version.as_u8();
```
Comment (sanket1729) [inline src/util/taproot.rs:703]: Here, we rely on Parity being only one bit. But converting to u8 should be enough.

**Review by Kixunil — APPROVED** [review]
ACK 91470f56c8f655454fdaa68826cdb886eac8a244

However, I'm highly surprised `Parity` is not an enum (`Even`, `Odd`) and even more strangely, `impl From<i32> for Parity` exists. I'd expect it to be `TryFrom`. I understand this is something from secp256k1, that's why I'm ACKing this even though I find it suspicious, maybe even wrong.

If someone could explain why it's not an enum, it'd be appreciated.

**Comment by sanket1729** [comment]
> If someone could explain why it's not an enum, it'd be appreciated.

This was a mistake in rust-secp and I the maintainers agree with it. I think we have to live with this now till the next release :(. 

For what it's worth, rust-bitcoin only has states Parity(1)/Parity(0) and there is no function that takes input Parity from the user.

**Comment by Kixunil** [comment]
As far as I can see, the consumer of rust-bitcoin could construct invalid `ControlBlock`. I personally wouldn't mind if fixing breaking release of secp256k1 was made (soonish) but if people feel it's not a good idea I can accept the sad reality.

**Review by dr-orlovsky — APPROVED** [review]
utACK 91470f56c8f655454fdaa68826cdb886eac8a244 in order to unlock the rest of PRs required for Taproot, which depends on this.

However I feel that we need another major `secp256k1` release fixing `Parity` (making it enum) before releasing new `rust-bitcoin` version.

**Comment by dr-orlovsky** [comment]
One use of deprecated function:
```
warning: use of deprecated associated function `secp256k1::ecdsa::recovery::<impl secp256k1::Secp256k1<C>>::sign_recoverable`: Use sign_ecdsa_recoverable instead.
   --> src/util/misc.rs:323:29
```
Since this is a non-critical warning I will merge this anyway, but it may be nice to fix it in some other PR

**Pull request merged**